File: VB/WpfApplication2/HeaderItemEx.vb
Imports System
Imports System.Globalization
Imports System.Linq
Imports System.Windows
Imports System.Windows.Data
Imports System.Windows.Markup
Imports DevExpress.Xpf.Spreadsheet
Imports DevExpress.Xpf.Spreadsheet.Internal
Imports System.Collections.ObjectModel
Imports System.Collections.Generic
Imports System.ComponentModel

Namespace WpfApplication2
	Public Class HeaderConverter
		Implements IMultiValueConverter

		Public Function Convert(ByVal value() As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
			If value Is Nothing OrElse value(0) Is Nothing Then
				Return Nothing
			End If
			Dim columnHeader As String = value(0).ToString()
			Dim collection As ObservableCollection(Of CustomHeaderCaption) = TryCast(value(1), ObservableCollection(Of CustomHeaderCaption))
			If collection Is Nothing OrElse collection.Count = 0 Then
				Return columnHeader
			End If

			Dim result As String = collection.Where(Function(l) l.OriginalCaption = columnHeader).Select(Function(caption) caption.NewHeader).FirstOrDefault()
			If result IsNot Nothing Then
				Return result
			End If

			Return columnHeader
		End Function

		Public Function ConvertBack(ByVal value As Object, ByVal targetTypes() As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
			Return TryCast(value, Object())
		End Function
	End Class
End Namespace

File: CS/WpfApplication2/HeaderItemEx.cs
using System;
using System.Globalization;
using System.Linq;
using System.Windows;
using System.Windows.Data;
using System.Windows.Markup;
using DevExpress.Xpf.Spreadsheet;
using DevExpress.Xpf.Spreadsheet.Internal;
using System.Collections.ObjectModel;
using System.Collections.Generic;
using System.ComponentModel;

namespace WpfApplication2
{
    public class HeaderConverter : IMultiValueConverter
    {
        public object Convert(object[] value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value == null || value[0] == null)
                return null;
            string columnHeader = value[0].ToString();
            ObservableCollection<CustomHeaderCaption> collection = value[1] as ObservableCollection<CustomHeaderCaption>;
            if (collection == null || collection.Count == 0)
                return columnHeader;

            string result = collection.Where(l => l.OriginalCaption == columnHeader).Select(caption => caption.NewHeader).FirstOrDefault();
            if (result != null)
                return result;

            return columnHeader;
        }

        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture)
        {
            return null;
        }
    }
}
